### PR TITLE
ekf2 support SET_GPS_GLOBAL_ORIGIN and purge globallocal usage

### PR DESCRIPTION
**Pull request by dagar**
- ~~requires https://github.com/PX4/PX4-ECL/pull/962~~
 - ekf2 and lpe supported
 - purge globallocalconverter usage (hidden global variables)
 - currently ekf2 doesn't acknowledge command to accomodate multi-EKF

### CODE REVIEW

File: src/modules/mavlink/mavlink_main.cpp [hunk 1]
```diff
@@ -1578,6 +1578,7 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
 		configure_stream_local("EXTENDED_SYS_STATE", 1.0f);
 		configure_stream_local("GLOBAL_POSITION_INT", 5.0f);
 		configure_stream_local("GPS2_RAW", 1.0f);
+		configure_stream_local("GPS_GLOBAL_ORIGIN", 0.1f);
```
Comment (mcsauder) [inline src/modules/mavlink/mavlink_main.cpp:1581]: Hi @dagar , did you want this value to be `1.0f` like the others?

File: src/modules/simulator/simulator_mavlink.cpp [hunk 2]
```diff
@@ -445,44 +445,40 @@ void Simulator::handle_message_hil_state_quaternion(const mavlink_message_t *msg
...
 		matrix::Eulerf euler = matrix::Quatf(hil_attitude.q);
 		hil_lpos.heading = euler.psi();
 		hil_lpos.xy_global = true;
 		hil_lpos.z_global = true;
-		hil_lpos.ref_lat = _hil_ref_lat;
-		hil_lpos.ref_lon = _hil_ref_lon;
-		hil_lpos.ref_alt = _hil_ref_alt;
-		hil_lpos.ref_timestamp = _hil_ref_timestamp;
+		hil_lpos.ref_timestamp = _global_local_proj_ref.timestamp;
+		hil_lpos.ref_lat = math::radians(_global_local_proj_ref.lat_rad);
```
Comment (mcsauder) [inline src/modules/simulator/simulator_mavlink.cpp:479]: Hi @dagar , did you want to use math::degrees() here instead?  Given the [struct definition](https://github.com/PX4/PX4-ECL/blob/99575d73e10aad06eb7d5b6415ea2db4151c5045/geo/geo.h#L74), I think that's what you were going for.  Let me know if I'm confused.

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/mavlink/mavlink_main.cpp:1581]: This message is a bit weird, the mavlink spec says after a new `SET_GPS_GLOBAL_ORIGIN` is received you're supposed to send `GPS_GLOBAL_ORIGIN`. So the idea is to stream by default at an extremely minimal rate, but if `SET_GPS_GLOBAL_ORIGIN` is received another `GPS_GLOBAL_ORIGIN` will go out immediately.

http://mavlink.io/en/messages/common.html#SET_GPS_GLOBAL_ORIGIN

File: src/modules/simulator/simulator_mavlink.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/simulator/simulator_mavlink.cpp:479]: Right!

**Review by mcsauder — APPROVED** [review]
@dagar , this looks good to go after you can update the changes in the `flight_task` files to the `flight_mode_manage` directory.  I'm setup to test if you can make that change.

**Comment by mcsauder** [comment]
Hi @dagar , failing test is hardware:
image
remaining work left to do in https://github.com/PX4/PX4-ECL/pull/962

**Comment by mcsauder** [comment]
As far as functionality goes things are lining up.... except for alt without GPS lock with HIL...
```
nsh> commander origin 45.000000001 111.000000001 150.125
nsh> Get NED origin (LLA):	45.0000000010, 111.0000000010, 0.000

```

**Comment by mcsauder** [comment]
It works!
```ekf2 status
ekf2 instance:		0
attitude valid:		true
local position valid:	true
global position (LLA):	40.5841001, -111.8316491, 1473.5
ekf2: ECL update: 2069 events, 14463us elapsed, 6.99us avg, min 1us max 75us 5.133us rms
ekf2: ECL full update: 1973 events, 550771us elapsed, 279.15us avg, min 155us max 1264us 109.347us rms
nsh> 
nsh> commander origin 45.0000001 111.0000000001 150.125
nsh> ecl: new vert pos: -1323.379272
nsh> Get NED origin (LLA): 45.0000001000, 111.0000000001, 150.125
```

**Comment by mcsauder** [comment]
@dagar , Everything checks out as much as I can test.  I think we're in good shape here.

File: src/modules/commander/Commander.cpp [hunk 3]
```diff
@@ -1196,37 +1216,39 @@ Commander::handle_command(const vehicle_command_s &cmd)
 			break;
 		}
 
-	case vehicle_command_s::VEHICLE_CMD_START_RX_PAIR:
```
Comment (mcsauder) [inline src/modules/commander/Commander.cpp:1199]: just alphabetized this list along with the addition of `vehicle_command_s::VEHICLE_CMD_SET_GPS_GLOBAL_ORIGIN`

File: msg/vehicle_command.msg [hunk 4]
```diff
@@ -82,6 +82,12 @@ uint16 VEHICLE_CMD_PAYLOAD_PREPARE_DEPLOY = 30001	# Prepare a payload deployment
 uint16 VEHICLE_CMD_PAYLOAD_CONTROL_DEPLOY = 30002	# Control a pre-programmed payload deployment
 uint16 VEHICLE_CMD_FIXED_MAG_CAL_YAW = 42006            # Magnetometer calibration based on provided known yaw. This allows for fast calibration using WMM field tables in the vehicle, given only the known yaw of the vehicle. If Latitude and longitude are both zero then use the current vehicle location.
 
+
+# PX4 vehicle commands (beyond mavlink commands)
+uint32 VEHICLE_CMD_SET_GPS_GLOBAL_ORIGIN = 65536        # Sets the GPS co-ordinates of the vehicle local origin (0,0,0) position. |Empty|Empty|Empty|Empty|Latitude|Longitude|Altitude|
```
Comment (dagar) [inline msg/vehicle_command.msg:87]: TODO: we need to add a check for this in mavlink so these commands aren't forwarded.

File: src/modules/commander/Commander.cpp [hunk 5]
```diff
@@ -359,6 +360,25 @@ int Commander::custom_command(int argc, char *argv[])
 		return (ret ? 0 : 1);
 	}
 
+	if (!strcmp(argv[0], "origin")) {
```
Comment (dagar) [inline src/modules/commander/Commander.cpp:363]: Maybe "set origin" or "set global origin"?

Comment (mcsauder) [reply src/modules/commander/Commander.cpp]: I settled on `set_ekf_origin`, let me know if you'd prefer it to be either of your suggestions, or feel free to just change it to anything you'd like!

**Comment by dagar** [comment]
Let's get the ECL side (https://github.com/PX4/PX4-ECL/pull/962) then cleanup the history in this branch.

File: src/modules/commander/Commander.cpp (hunk 5 above)
Comment (dagar) [reply src/modules/commander/Commander.cpp]: I think that works and might help reduce confusion with the home position.

**Comment by mcsauder** [comment]
Rebased against current PX4 master.

File: msg/vehicle_command.msg (hunk 4 above)
Comment (dagar) [reply msg/vehicle_command.msg]: Done.

**Comment by dagar** [comment]
Due to minor API changes this PR is blocking further ECL updates.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 6]
```diff
@@ -2680,31 +2686,29 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
...
 
 		vehicle_local_position_s hil_local_pos{};
 		hil_local_pos.timestamp = timestamp;
 
-		hil_local_pos.ref_timestamp = _hil_local_proj_ref.timestamp;
-		hil_local_pos.ref_lat = math::radians(_hil_local_proj_ref.lat_rad);
-		hil_local_pos.ref_lon = math::radians(_hil_local_proj_ref.lon_rad);
-		hil_local_pos.ref_alt = _hil_local_alt0;
+		hil_local_pos.ref_timestamp = _global_local_proj_ref.timestamp;
+		hil_local_pos.ref_lat = math::radians(_global_local_proj_ref.lat_rad);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:2702]:
```suggestion
		hil_local_pos.ref_lat = math::degrees(_global_local_proj_ref.lat_rad);
```

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 7]
```diff
@@ -2680,31 +2686,29 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
...
+		hil_local_pos.ref_lon = math::radians(_global_local_proj_ref.lon_rad);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:2703]:
```suggestion
		hil_local_pos.ref_lon = math::degrees(_global_local_proj_ref.lon_rad);
```

**Comment by dagar** [comment]
Rebased on current master.

**Review by mcsauder — APPROVED** [review]
This is working for me and everything looks very good.  There are a few moving parts we need to contend with, but you've done very, very great work here @dagar.

**Review by mcsauder — APPROVED** [review]
I haven't flight tested this new functionality yet but will soon.  Aside from that, everything I can check checks out.  Nice work @dagar !

**Comment by mrpollo** [comment]
@dagar is it helpful if the community pitches in with a couple of flight tests? if so, can you give us a few requirements in terms of the hardware setup and the testing procedures

**Comment by mcsauder** [comment]
> @dagar is it helpful if the community pitches in with a couple of flight tests? if so, can you give us a few requirements in terms of the hardware setup and the testing procedures

I'll make that a priority, my apologies that I haven't gotten flight data collected yet!

**Comment by dagar** [comment]
Fixing merge conflicts after https://github.com/PX4/PX4-Autopilot/pull/16739 went into master.

**Pull request merged**